source/common/upstream/subset_lb.cc

```diff
@@ -509,52 +509,43 @@ SubsetLoadBalancer::PrioritySubsetImpl::PrioritySubsetImpl(const SubsetLoadBalan
 void SubsetLoadBalancer::HostSubsetImpl::update(const HostVector& hosts_added,
                                                 const HostVector& hosts_removed,
                                                 std::function<bool(const Host&)> predicate) {
-  std::unordered_set<HostSharedPtr> predicate_added;
+  HostVectorSharedPtr hosts(new HostVector());
+  HostVectorSharedPtr healthy_hosts(new HostVector());
+  HostVectorSharedPtr degraded_hosts(new HostVector());
 
-  HostVector filtered_added;
-  for (const auto host : hosts_added) {
+  // We cache the result of matching the host against the predicate. This ensures
+  // that we maintain a consistent view of the metadata and saves on computation
+  // since metadata lookups can be expensive.
+  //
+  // We use an unordered_set because this can potentially be in the tens of thousands.
+  std::unordered_set<const Host*> matching_hosts;
+
+  auto cached_predicate = [&matching_hosts](const auto& host) {
+    return matching_hosts.count(&host) == 1;
+  };
+
+  // TODO(snowp): If we had a unhealthyHosts() function we could avoid potentially traversing
+  // the list of hosts twice.
+  for (const auto host : original_host_set_.hosts()) {
```

Comment:
isn't `const auto& host` cheaper here? [avoid copying the `shared_ptr`, which implies increasing refcount, etc]